### ISSUE TITLE
Suggestion: NeverPipeOnce

### ISSUE THREAD

**Issue by josevalim** [rrrene/credo#221]
We at the Elixir core team consider a smell to use the pipe just once: `value |> Some.fun`. The pipe was designed to express pipelines and you don't have a pipeline if you have only one call. :) Therefore it would be nice to have a check that guarantees you never pipe only once.

PS: We are adding Credo to Ecto via ebert: https://ebertapp.io/github/elixir-ecto/ecto/reviews/1. It is working quite well so far! Thank you!

PS: Read the discussion below for examples of where piping once is actually useful in case you want to disable the check.

**Comment by rrrene** [comment]
It is a good idea to have a check for this. :+1: 

I did not know of this preference by the core team. Just as an anecdote: The "single pipe" helped me alot in the beginning because `list |> Enum.take(5)` almost felt like `list.take(5)` (coming from an OO background).

**Comment by josevalim** [comment]
That's part of why we believe it is an important check. It helps to keep old habits at home. :)

**Comment by rrrene** [comment]
> It helps to keep old habits at home. :)

Haha, I might just print & frame that. :grin:

**Comment by sstern6** [comment]
@rrrene would you say this would be a `Refactor` check or a `Warning`?  or neither 😄 ?

**Comment by svoynow** [comment]
I'm working on this one at the moment. I'm making it a `Refactor` for the moment, but I'm happy to change.

**Comment by forki** [comment]
But isn't a single pipe the beginning of writing a longer pipeline? I'm working > 10 years in F# and I think this really artificial constraint. Are there any technical limitations that lead to this suggestion?

**Comment by josevalim** [comment]
@forki I would consider this about enforcing a style rather than being technical limitation. For new developers, sometimes a single pipe is used to cope with the lack of OO methods:

```elixir
def my_function(arg, msg) when arg |> is_tuple do
  arg |> send(msg |> format)
end
```

which in turns makes the function much harder to read.

I cannot think of any situation where a single pipe is more readable than no pipes so I would like to enforce that in my codebases but your mileage may vary. :)

**Comment by polytypic** [comment]
I have to second @forki, though I have only cursory interest in Elixir at this point. I also have a long history of programming and programming in functional languages.  I designed [A family of function application operators for Standard ML](http://comp.lang.functional.narkive.com/zZJZg20r/a-family-of-function-application-operators-for-standard-ml) some 11 years ago.

One problem with deprecating 1 element long "pipelines" is that it creates an artificial need to change the layout of the code as you need to move "the object of attention" (the initial value) from the end to the beginning or vice versa.  A change to add or remove an operation from a pipeline now becomes more complex.  Sometimes one wants to emphasize the object of attention even when only a single operation is being performed and the pipeline operator makes that possible.

Of course, like pretty much all features of programming languages, it is also possible to abuse the pipeline operator as you show.

**Comment by josevalim** [comment]
@forki @polytypic I completely understand those arguments and I agree those are valid use cases. Unfortunately those are not the main reasons I see it being use in practice so I think a nudge to the right direction can be helpful. Especially because, if you always pipe everything, than the use case of emphasizing one particular value becomes moot.

> One problem with deprecating 1 element long "pipelines" 

I just want to make absolutely clear this is not a discussion about language features, this is about using Credo which is completely separate from the Elixir language. Credo allows you to enforce rules which you can opt-in and opt-out as you prefer. There is nothing being deprecated.

If your team is more comfortable with functional aspects and you would like to use the pipeline on whatever way you wish then you can simply disable the warning on Credo. I do disable a couple checks on Credo whenever I use it. Elixir will not change based on this.

**Comment by forki** [comment]
@polytypic thanks. Excatly refactoring will become awkward and I wonder what happens if such a linter would work while I'm wrting code. It would just constantly complain during my writing.

Also I often write pipes (and yes even single pipes) to work easier with a REPL.

  x
  |> doSomething

this allows me to select x and evaluate it. Then I select one line more and evaluate again. So I see input and output

**Comment by rrrene** [comment]
Since Credo is designed to be an educational tool which helps people develop their skills, I wonder if we can convey these considerations in the explanation texts (we are [discussing these already](https://github.com/rrrene/credo/pull/240#pullrequestreview-10021793)). 

As for advanced usage: people are always free to assign a lower priority to this check so it does not appear in the regular report or deactivate it completely.

It is an important part of each talk I give on Credo to emphasize that this is just a tool. As you improve your functional programming skills, you certainly want to tweak the default settings/checks to your liking :grin:

**Comment by forki** [comment]
> you certainly want to tweak the default settings/checks to your liking 

sure, but:

> "We at the Elixir core team consider a smell to use the pipe just once"

So this is a really strong statement and it will influence the way how elixir will be written in the wild. Even without a Linter people will use this statement to in code reviews to criticize style.

**Comment by josevalim** [comment]
@forki it is no surprise the Elixir team has a opinion. :) We also have a strong opinion on macros, mocks, on function naming, etc.

> Even without a Linter people will use this statement to in code reviews to criticize style.

It is only a problem if whatever the Elixir team says is used blindly (which is hardly our fault). A team should be able to discuss such statements and evaluate if it applies to them. Luckily this issue itself is a good starting point.

**Comment by forki** [comment]
@josevalim absolutely. Elixir team should have a strong opinion.

**Comment by josevalim** [comment]
I have edited the original issue to mention there is a discussion down here with reasons do disable the check.

**Comment by rrrene** [comment]
Împlemented in #240.
